In [9]:
import timeit, time

# Selection Sort

Percorre a lista de elementos e seleciona o menor elemento a cada passo. E em seguida, troca esse elemento com o primeiro não ordenado.

In [2]:
def select_sort(v, n):
    
    for i in range(n):
        min_index = i
        for j in range(i + 1, n):
            if v[j] < v[min_index]:
                min_index = j

        v[i], v[min_index] = v[min_index], v[i]

In [3]:
# Exemplo de uso:
vetor = [64, 25, 12, 22, 11]
tamanho = len(vetor)

select_sort(vetor, tamanho)
print("Vetor ordenado:", vetor)

Vetor ordenado: [11, 12, 22, 25, 64]


#### Exemplo de algortimo e uso com os arquivos do sigaa

In [19]:
def select_sort(v):
    
    n = len(v)
    for i in range(n):
        min_index = i
        for j in range(i + 1, n):
            if v[j] < v[min_index]:
                min_index = j
        v[i], v[min_index] = v[min_index], v[i]

def main(input_file, output_file):
    
    try:
        
        with open(input_file, 'r') as file:
            lines = file.readlines()

        values = [int(line.strip()) for line in lines]

        start_time = time.time() 
        select_sort(values)
        end_time = time.time()

        with open(output_file, 'w') as file:
            for value in values:
                file.write(str(value) + '\n')

        print(f"Arquivo ordenado com sucesso. Resultados salvos em {output_file}")
        
        execution_time = end_time - start_time
        print(f"Tempo de execução: {execution_time} segundos")

    except Exception as e:
        print(f"Ocorreu um erro: {e}")

if __name__ == "__main__":
    
    input_file = 'instancias-numericas/instancias-num/num.1000.4.in'
    output_file = "output_select.in" 
    main(input_file, output_file)

Arquivo ordenado com sucesso. Resultados salvos em output_select.in
Tempo de execução: 0.15765929222106934 segundos


**Complexidade**: O(n²), pois o algoritmo realiza uma busca no restante do vetor para encontrar o menor elemento e troca o elemento atual com o menor elemento encontrado. Logo, usamos dois loops aninhados, um dentro do outro.

#  Insertion sort

Ordena a lista em ordem crescente, movendo elementos maiores para a esquerda.

In [1]:
def insertion_sort(v, n):
    
    for i in range(1, n):
        chave = v[i]
        j = i - 1
        while j >= 0 and chave < v[j]:
            v[j + 1] = v[j]
            j -= 1
        v[j + 1] = chave

In [2]:
# Exemplo de uso:
vetor = [9, 2, 5, 1, 7, 6, 3]
tamanho = len(vetor)

insertion_sort(vetor, tamanho)
print("Vetor ordenado:", vetor)

Vetor ordenado: [1, 2, 3, 5, 6, 7, 9]


#### Exemplo de algortimo e uso com os arquivos do sigaa

In [3]:
def main(input_file, output_file):
    
    try:
        
        with open(input_file, 'r') as f:
            lines = f.readlines()
            n = int(lines[0])
            v = list(map(int, lines[1:]))
            
        start_time = time.time() 
        insertion_sort(v, n)
        end_time = time.time() 

        with open(output_file, 'w') as f:
            f.write(str(n) + '\n')
            for value in v:
                f.write(str(value) + '\n')
                

        print(f"Arquivo ordenado com sucesso. Resultados salvos em {output_file}")
        
        execution_time = end_time - start_time
        print(f"Tempo de execução: {execution_time} segundos")
        
    except Exception as e:
        print(f"Ocorreu um erro: {e}")

if __name__ == "__main__":
    
    input_file = 'instancias-numericas/instancias-num/num.1000.1.in'
    output_file = "output_insertion.in" 
    main(input_file, output_file)

Ocorreu um erro: name 'time' is not defined


**Complexidade**: O(n²), pois, em cada iteração, ele percorre todos os elementos anteriores ao elemento atual, o que resulta em uma soma de 1 + 2 + 3 + ... + (n-1) comparações e movimentos de elementos, ou seja, uma P.A.

# Merge Sort

Divide a lista em pequenas partes, ordena essas partes e, em seguida, combina-as para produzir a lista ordenada final. 

In [29]:
def merge(v, i, meio, f):
    
    n1 = meio - i + 1
    n2 = f - meio

    P1 = [0] * n1
    P2 = [0] * n2

    for k in range(n1):
        P1[k] = v[i + k]
    for j in range(n2):
        P2[j] = v[meio + 1 + j]

    i_temp, j_temp, k = 0, 0, i
    while i_temp < n1 and j_temp < n2:
        
        if P1[i_temp] <= P2[j_temp]:
            v[k] = P1[i_temp]
            i_temp += 1
            
        else:
            v[k] = P2[j_temp]
            j_temp += 1
        k += 1

    # Copiar os elementos restantes de P1[], se houver algum
    while i_temp < n1:
        v[k] = P1[i_temp]
        i_temp += 1
        k += 1

    # Copiar os elementos restantes de P2[], se houver algum
    while j_temp < n2:
        v[k] = P2[j_temp]
        j_temp += 1
        k += 1
        
def merge_sort(v, i, f):
        
    if i < f:
        meio = (i + f) // 2
        merge_sort(v, i, meio)
        merge_sort(v, meio + 1, f)
        merge(v, i, meio, f)

In [30]:
# Exemplo de uso
vetor = [12, 11, 13, 5, 6, 7]
#f = len(vetor) - 1

merge_sort(vetor, 0, len(vetor) - 1)
print("Vetor ordenado:", vetor)

Vetor ordenado: [5, 6, 7, 11, 12, 13]


#### Exemplo de algortimo e uso com os arquivos do sigaa

In [36]:
def merge_sort_file(input_file, output_file):
    
    try:
        with open(input_file, 'r') as file:
            lines = file.readlines()
            vetor = [int(line.strip()) for line in lines]

        start_time = time.time()
        merge_sort(vetor, 0, len(vetor) - 1)
        end_time = time.time()

        with open(output_file, 'w') as file:
            for num in vetor:
                file.write(str(num) + '\n')

        execution_time = end_time - start_time
        print(f"Tempo de execução: {execution_time} segundos")
        
    except Exception as e:
        print(f"Ocorreu um erro: {e}")

if __name__ == "__main__":
    
    input_filename = "instancias-numericas/instancias-num/num.1000.3.in"
    output_filename = "output_merge.in"
    
    merge_sort_file(input_filename, output_filename)
    print("Arquivo ordenado com sucesso. Resultados salvos em", output_filename)

Tempo de execução: 0.0026187896728515625 segundos
Arquivo ordenado com sucesso. Resultados salvos em output_merge.in


**Complexidade** : O(n log n) pois, nesse caso, dividimos para conquistar. 

O algoritmo divide repetidamente o vetor em duas metades e combina essas metades de maneira ordenada. A divisão ocorre até que o tamanho do subvetor seja reduzido a 1, o que é um processo que requer log2(n) etapas. Em cada etapa, ocorre uma junção das partes, resultando em um tempo total de O(n log n).

## Comparação do tempo de execução dos algoritmos com o arquivo **'num.1000.3.in'** 

Select_sort : *0.15765929222106934* segundos <br>
Insert_sort : *0.1429891586303711* segundos <br>
Merge_sort :  *0.0026187896728515625* segundos

O Merge Sort é muito mais eficiente, especialmente para entradas maiores. Justamente pela sua forma de ordenação. (Divide para consquistar)